src/sink/reduce.ts

```diff
@@ -0,0 +1,47 @@
+// sink/reduce.ts
+
+/*
+ * Copyright (c) 2021 Check Digit, LLC
+ *
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+
+import asyncerator, { Asyncerator } from '../asyncerator';
+
+/**
+ * Calls the specified callback function for all the elements in a stream. The return value of the callback function
+ * is the accumulated result, and is provided as an argument in the next call to the callback function.
+ * Equivalent to the Javascript Array.reduce() method.
+ *
+ * @param reduceFunction The reduce method calls the reduceFunction function one time for each element in the stream.
+ * @param initialValue If initialValue is specified, it is used as the previousValue to start the accumulation.
+ * Otherwise, the initial previousValue will be undefined.
+ */
+
+export type ReduceFunction<Input, Output> = (
+  previousValue: Output,
+  currentValue: Input,
+  currentIndex: number
+) => Output;
+
+export default function <Input>(
+  reduceFunction: ReduceFunction<Input, Input>,
+  initialValue?: Input
+): (iterator: Asyncerator<Input>) => Promise<Input | undefined>;
+export default function <Input, Output>(
+  reduceFunction: ReduceFunction<Input, Output>,
+  initialValue?: Output
+): (iterator: Asyncerator<Input>) => Promise<Output | undefined>;
+export default function <Input, Output>(
```

Comment:
I don't really understand why we are using 'Input | Output | undefined' as the output type. To me it would seem that it should just be a simple 'Output | undefined'. Why do we need the Input?

Reply:
the first override uses just `Input | undefined`.  I basically matched the overrides that `Array.reduce` uses.

Reply:
I looked at the Array.reduce types and I see the following declarations:
    reduce(callbackfn: (previousValue: T, currentValue: T, currentIndex: number, array: T[]) => T): T;
    reduce(callbackfn: (previousValue: T, currentValue: T, currentIndex: number, array: T[]) => T, initialValue: T): T;
    reduce<U>(callbackfn: (previousValue: U, currentValue: T, currentIndex: number, array: T[]) => U, initialValue: U): U;

nothing that would return T | U ?

Reply:
the difference is that the actual implementation function declaration has to be a superset (or implement) all the various overloads.  Since the three declarations return either T or U, the typescript implementation (if there was one) would need to return T | U.

Reply:
Ok, I had a read on overload typing / implementation signatures from the typescript handbook. Didn't realize the implementation signature wouldn't be callable. lgtm.